File: mosaic.py

class Tile(object):
    
    def __init__(self, height, width, row=None, col=None, **kw):
        self.width = width
        self.height = height
        self.row = row
        self.col = col
        self.__dict__.update(kw)

    def copy(self):
        return self.__class__(**self.__dict__)


class Mosaic(object):

    def __init__(self, width):
        self._width = width
        self._holes = []
        self._tiles = {}

    @property
    def width(self):
        return self._width

    @property
    def height(self):
        return len(self._holes)

    def add(self, tile=None, **kw):
        tile = tile.copy() or Tile(**kw)
        if not (tile.row and tile.col):
            self._find_hole(tile)
        self._occupy(tile)
        return tile

    def linearize(self):
        seen = set()
        for r in xrange(len(self._holes)):
            for c in xrange(self._width):
                t = self._tiles.get((r, c))
                if not t:
                    continue
                if t not in seen:
                    yield t
                    seen.add(t)

    def _find_hole(self, tile):
        for r in xrange(len(self._holes)):
            for c in xrange(self._width):
                if self._holes[r] and self._holes[r][c]:
                    if self._fits_holes(tile, r, c):
                        tile.row = r
                        tile.col = c
                        return tile
        tile.row = len(self._holes)
        tile.col = 0
        return tile

    def _fits_holes(self, tile, r, c):
        for R in xrange(r, r + tile.height):
            for C in xrange(c, c + tile.width):
                if C >= self._width:
                    return False
                if R < len(self._holes) and (not self._holes[R] or not self._holes[R][C]):
                    return False
        return True

    def _occupy(self, tile):
        while len(self._holes) < tile.row + tile.height:
            self._holes.append([True] * self._width)
        for r in xrange(tile.row, tile.row + tile.height):
            for c in xrange(tile.col, tile.col + tile.width):
                self._holes[r][c] = False
                self._tiles[(r,c)] = tile
            if not any(self._holes[r]):
                self._holes[r] = False

File: mosaic.js

var Mosaic = function(width) {
    this.width = width;
    this.height = 0;
    this.tiles = []
    this.holes = []
}


Mosaic.prototype.place = function(obj) {
    this.tiles.push(obj);
    this.occupy(obj);
}


Mosaic.prototype.append = function(w, h) {

    var obj = this.find(w, h);
    obj.width = w;
    obj.height = h;

    this.occupy(obj);

    obj.index = this.tiles.length;
    this.tiles.push(obj)

    return obj;
}


Mosaic.prototype.find = function(w, h) {

    for (var r = 0; r < this.holes.length; r++) {
        for (var c = 0; c < this.width; c++) {
            if (this.holes[r] && this.holes[r][c]) {
                if (this._fits(w, h, r, c)) {
                    return {row: r, col: c};
                }
            }
        }
    }

    return {row: this.holes.length, col: 0};
}


Mosaic.prototype._fits = function(w, h, r, c) {
    // console.log('fits', {width: w, height: h, row: r, col: c});
    for (var R = r; R < r + h; R++) {
        for (var C = c; C < c + w; C++) {
            // console.log('fits', {width: w, height: h, row:r, col:c, R:R, C:C, hole:this.holes[R] && !this.holes[R][C]})
            if (R < this.holes.length && (!this.holes[R] || !this.holes[R][C])) {
                return false;
            }
        }
    }
    return true;
}


Mosaic.prototype.occupy = function(obj) {

    while (this.holes.length < obj.row + obj.height + 2) {
        var row = [];
        for (var i = 0; i < this.width; i++) {
            row.push(true);
        }
        this.holes.push(row);
    }

    for (var r = obj.row; r < obj.row + obj.height; r++) {
        
        if (!this.holes[r]) {
            throw "invalid row " + r;
        }

        for (var c = obj.col; c < obj.col + obj.width; c++) {
            this.holes[r][c] = false;
        }

        var empty = true;
        for (var c = 0; c < this.width; c++) {
            if (this.holes[r][c]) {
                empty = false;
                break;
            }
        }
        if (empty) {
            this.holes[r] = null;
        }

    }


}


if (module) {
    module.exports = {
        Mosaic: Mosaic
    }
}
